frontend/src/components/pages/TranslationPage.tsx

```diff
@@ -100,6 +124,63 @@ const TranslationPage = () => {
     );
   };
 
+  const onUserInput = async (newContent: string, lineIndex: number) => {
+    const oldContent =
+      translatedStoryLines[arrayIndex(lineIndex)].translatedContent;
+    if (oldContent !== undefined) {
+      updateVersionHistory(oldContent, lineIndex);
+    }
+    onChangeTranslationContent(newContent, lineIndex);
+  };
+
+  const undoChange = () => {
+    if (versionHistoryStack.Undo.length > 0) {
+      const { lineIndex, content: newContent } = versionHistoryStack.Undo[
+        versionHistoryStack.Undo.length - 1
+      ];
+      const oldContent =
+        translatedStoryLines[arrayIndex(lineIndex)].translatedContent;
+      if (oldContent !== newContent) {
+        const newHistory = {
+          Undo: deepCopy(versionHistoryStack.Undo.slice(0, -1)),
```

Comment:
Might be a slow night for me; could you explain why we need the `deepCopy`?

Reply:
Yeah no worries, this took me forever to figure out. Basically without deepCopy, Typescript thinks its smart and stores a reference to the object rather than the actual contents, but that means that every object in the stack will update every time a new element is added.